### PR TITLE
Improve calculations of ETH spent metrics

### PR DESCRIPTION
**Pull request by IvanIvanoff**
#### Summary
Still in work.

[//]: # (#### Related PRs)

[//]: # (#### Additional deploy notes)

[//]: # (#### Screenshots)

### CODE REVIEW

File: lib/sanbase/utils/application_utils.ex [hunk 1]
```diff
@@ -0,0 +1,29 @@
+defmodule Sanbase.ApplicationUtils do
+  @doc ~s"""
+  Start a worker/supervisor only in particular environment(s).
+  Example: Not startuing `MySupervisor` in tests can now be done by replacing
+  `{MySupervisor, []}` in the supervisor children by
+  `run_in({MySupervisor, []}, [:dev, :prod])`
```
Comment (tspenov) [inline lib/sanbase/utils/application_utils.ex]: `start_in` instead of `run_in` ?

File: lib/sanbase/utils/application_utils.ex [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  INPORTANT NOTE: If you use it, you must use `normalize_children` on the children list.
+  """
+  defmacro start_in(expr, environments) do
```
Comment (tspenov) [inline lib/sanbase/utils/application_utils.ex]: why it should be macro ? Could it be a regular public function ?

File: lib/sanbase_web/graphql/resolvers/project/project_resolver.ex [hunk 3]
```diff
@@ -114,6 +112,8 @@ defmodule SanbaseWeb.Graphql.Resolvers.ProjectResolver do
     {:ok, projects}
   end
 
+  # Helper functions
```
Comment (tspenov) [inline lib/sanbase_web/graphql/resolvers/project/project_resolver.ex]: Why is the comment moved here ? I think right below is a public resolver function.

File: lib/sanbase_web/graphql/resolvers/project/project_transactions_resolver.ex [hunk 4]
```diff
@@ -138,22 +141,40 @@ defmodule SanbaseWeb.Graphql.Resolvers.ProjectTransactionsResolver do
 
         {:ok, []}
     end
+  rescue
```
Comment (tspenov) [inline lib/sanbase_web/graphql/resolvers/project/project_transactions_resolver.ex]: Is it necessary to use rescue. Isn't connection errors catched by clickhousex or clickhouse_ecto ?

File: lib/sanbase/utils/application_utils.ex (hunk 2 above)
Comment (IvanIvanoff) [reply lib/sanbase/utils/application_utils.ex]: it does not to be a macro

File: lib/sanbase_web/graphql/resolvers/project/project_resolver.ex (hunk 3 above)
Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/resolvers/project/project_resolver.ex]: Will rearrange them

File: lib/sanbase/utils/application_utils.ex (hunk 2 above)
Comment (tspenov) [reply lib/sanbase/utils/application_utils.ex]: ok, so then can we simplify it by using a function

File: lib/sanbase/utils/application_utils.ex [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  @doc ~s"""
+  Start a worker/supervisor only in particular environment(s).
+  Example: Not startuing `MySupervisor` in tests can now be done by replacing
+  `{MySupervisor, []}` in the supervisor children by
+  `start_in({MySupervisor, []}, [:dev, :prod])`
+
+  INPORTANT NOTE: If you use it, you must use `normalize_children` on the children list.
+  """
+  def start_in(expr, environments) do
+    require Sanbase.Utils.Config, as: Config
```
Comment (tspenov) [inline lib/sanbase/utils/application_utils.ex]: Can we move this require out of the function ?

File: lib/sanbase_web/graphql/resolvers/project/project_transactions_resolver.ex [hunk 6]
```diff
@@ -51,127 +47,85 @@ defmodule SanbaseWeb.Graphql.Resolvers.ProjectTransactionsResolver do
...
   """
   def eth_spent_by_erc20_projects(_, %{from: from, to: to}, _resolution) do
-    with projects when is_list(projects) <- Project.erc20_projects(),
-         {:ok, total_eth_spent} <-
-           Clickhouse.EthTransfers.eth_spent_by_projects(projects, from, to) do
+    with projects when is_list(projects) <- Project.erc20_projects() do
+      total_eth_spent =
+        projects
+        |> Enum.map(&Task.async(fn -> calculate_eth_spent_cached(&1, from, to).() end))
+        |> Enum.map(&Task.await/1)
```
Comment (tspenov) [inline lib/sanbase_web/graphql/resolvers/project/project_transactions_resolver.ex]: we could probably extract this into some parallel map func into utils that takes collection and function ?

**Pull request merged**